feat(peakfinder): add algorithm for finding a peak in a 1d list

mattfreer opened the pull request:
This PR adds a Peakfinder algorithm example.

A peak is any element in a list that is larger or equal to both it's neighbours
Items that only have one neighbour (end items) are considered a peak if
they are larger or equal to their neighbour.

mattfreer committed:
feat(peakfinder): add algorithm for finding a peak in a 1d list

A peak is any element that is larger or equal to both it's neighbours
Items that only have one neighbour (end items) are considered a peak if
they are larger or equal to their neighbour.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+  (:use peakfinder))
+
+(deftest finding-a-peak
+  (is (contains? (set '(7)) (find-peak '(3 2 4 7 5 6 2))))
```
You can use the set literal syntax here.
(set '(3 5 2)) is the same as #{3 5 2}.

Also, you should almost always use vector literals rather than quoted lists when you have a literal sequence.  (of course in this case you should have used the set literal as explained above).  So instead of '(5 4 2 2) you should use [5 4 2 2].  They are both seqs and can be consumed by any of the functions in the seq library.  Vectors and Lists do have different performance characteristics which you should look up and become familiar with.  Also, the vector and list versions differ in this important way.  Each item in the vector version is evaluated and put into the vector.  The quoted list version quotes the whole thing so items are not evaluated.  In the above example that makes no difference, but here is an example where it does matter.

[(+ 1 2) (+ 2 3)] => [3 5]

While
 '((+ 1 2) (+ 2 3))  => ((+ 1 2) (+ 2 3))

The quoted list just evaluates to exactly itself.  That's what quoting does.  So the resulting list contains the two items.  The first item is a list whose contents are 1) the symbol "+"  2) the number 1 and 3) the number 2.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+  ([list] (->> (middle-val list) (item list) (find-peak list)))
+
+  ([list current-item]
+   (if (empty? list) -1
```
For some reason, it is considered more "idiomatic" to check the emptiness of a seq with (seq list) rather than (empty? list).  Note they return opposite values!  (seq list) ~= (not (empty? list))

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+
+(ns peakfinder)
+
+(defn ^:dynamic find-peak
```
There is really no reason to make this ^:dynamic.  You really only need dynamic if you intend to rebind it dynamically in a (rebind [...]) form.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+  "returns a map that represents an item in a list. The data describes the
+  index at which the item resides in the list and it's value. e.g. {index 1 :value 3}"
+  [list, index]
+  {:index index :value (nth list index nil)})
```
This is potentially dangerous performance wise if list is a list or a seq or something that does not have random access.  See comment about this below.  One general approach to solving this would be to convert your input list to a vector at the very beginning of the algorithm and ensure that you use vectors throughout.  Vectors support fast access to particular indexes.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+  "determines if an item is a peak compared to the supplied comparitors"
+  [item & comparitors]
+  (let [item-value (:value item)
+        values (map (fn [item] (:value item)) comparitors)]
```
`(fn [item] (:value item))` is a function which takes a map and returns the :value element in that map.  Note that keywords are, in addition to being string-like values, functions which take a map and return the value in that map corresponding to the key.  This is actually why `(:value item)` works.  The :value function looks up the value under the :value key in its argument.  It looks like some fancy syntactic sugar, but it's not.  :value is actually a function, so you can call it just like any other function - by putting it first in a list. e.g. `(:value map)`.  The reason I'm pointing all this out is that the 'map' function expects a function as its first value.  You have passed `(fn [item] (:value item))`.  This is exactly the same as just :value.  In other words, you could have done:
(map :value comparitors)

mattfreer-mdsol replied:
@kdavy Great explanation. It's really neat that Keywords are functions too.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,82 @@
+(defn- left-neighbour
+  "returns the left neighbour of an item in a list"
+  [list current-item]
+  (item list (- (:index current-item) 1)))
```
Clojure has inc and dec functions for adding/subtracting one from a value.

mattfreer committed:
feat(peakfinder): find all peaks in a sequence

Peakfinder has been re-written to find all peaks in a sequence. Now uses
core.match.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,54 @@
+(ns peakfinder
+  (:require [clojure.core.match :refer [match]]))
+
+(defun find-peak [input]
```
I think it would be better if this function threw an exception on an invalid input instead of returning nil.

kdavy commented on src/peak_finding/1d_list.clj:
```diff
@@ -0,0 +1,54 @@
+  "Returns all peaks in a sequence"
+  (match [input]
+         ;; checks input is a sequence of multiple numbers
+         [(input :guard #(and (sequential? %)
```
Instead of checking for (> (count %) > 1) you could just put the [([h] :seq)] case ahead of this case.
